# Segmenting and Clustering Neighborhoods in Toronto - Part 3

## Importing libraries

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

print("\nImport completed")

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.8.5
  latest version: 4.9.0

Please update conda by running

    $ conda update -n base -c defaults conda



# All requested packages already installed.

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.8.5
  latest version: 4.9.0

Please update conda by running

    $ conda update -n base -c defaults conda



# All requested packages already installed.


Import completed


## Download and Explore Dataset

Let's load the Geospatial dataset for Toronto Neighborhoods and see how many neighborhoods and boroughs we have.

In [2]:
df_toronto= pd.read_csv('Toronto_Geospatial.csv')

print('\nThe dataset has {} Boroughs and {} Neighborhoods\n'.format(len(df_toronto['Borough'].unique()),len(df_toronto['Neighborhood'])))
df_toronto.head()


The dataset has 10 Boroughs and 103 Neighborhoods



PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

#### Create a map of Toronto with neighborhoods superimposed on top.


In [3]:
#Use geopy library to get the latitude and longitude values of Toronto City

address = 'Toronto, Ontario'

# Define a user_agent names "toronto_explorer" to create an instance of the geocoder
geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('\n The geograpical coordinate of Toronto are {}, {}.\n'.format(latitude, longitude))


# Create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# Add markers to map
for lat, lng, borough, neighborhood in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Borough'], df_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto


 The geograpical coordinate of Toronto are 43.6534817, -79.3839347.



## Toronto Neighborhood Analysis

For our analisys we decide to work only with boroughs that contain the word Toronto. Let's create a new dataframe with these boroughs.

In [4]:
df_boroughs = df_toronto[df_toronto['Borough'].str.contains('Toronto')].reset_index(drop = True)
df_boroughs.head(10)

PostalCode          Borough  \
0        M4E     East Toronto   
1        M4K     East Toronto   
2        M4L     East Toronto   
3        M4M     East Toronto   
4        M4N  Central Toronto   
5        M4P  Central Toronto   
6        M4R  Central Toronto   
7        M4S  Central Toronto   
8        M4T  Central Toronto   
9        M4V  Central Toronto   

                                        Neighborhood   Latitude  Longitude  
0                                        The Beaches  43.676357 -79.293031  
1                       The Danforth West, Riverdale  43.679557 -79.352188  
2                     India Bazaar, The Beaches West  43.668999 -79.315572  
3                                    Studio District  43.659526 -79.340923  
4                                      Lawrence Park  43.728020 -79.388790  
5                                   Davisville North  43.712751 -79.390197  
6                  North Toronto West, Lawrence Park  43.715383 -79.405678  
7                                         Davisville  43.704324 -79.388790  
8                        Moore Park, Summerhill East  43.689574 -79.383160  
9  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049

Let's see how many boroughs and neighborhoods we obtained

In [5]:
print('\nThe dataset has {} Boroughs and {} Neighborhoods '.format(len(df_boroughs['Borough'].unique()),len(df_boroughs['Neighborhood'].unique())))


The dataset has 4 Boroughs and 39 Neighborhoods 


#### Let's plot the map and the markers for these neighborhoods

In [6]:
#Use geopy library to get the latitude and longitude values of Toronto City

address = 'Central Toronto, Toronto'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude

# Create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

# Add markers to map
for lat, lng, borough, neighborhood in zip(df_boroughs['Latitude'], df_boroughs['Longitude'], df_boroughs['Borough'], df_boroughs['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

## Explore the neighborhoods

As mentionned before there are 4 boroughs that contain the word 'Toronto' and a total of 39 neighborhoods. In the following we will use Foursquare API to explore venues.

In [7]:
print('\nNumber of neighborhoods for each boroughs: \n\n{} '.format(df_boroughs['Borough'].value_counts()))


Number of neighborhoods for each boroughs: 

Downtown Toronto    19
Central Toronto      9
West Toronto         6
East Toronto         5
Name: Borough, dtype: int64 


Let's define Foursquare Credentials and Version (*credentials have been changed for the privacy*) 

In [8]:
# @hidden_cell
CLIENT_ID = 'MYID'      
CLIENT_SECRET = 'MYOWD'   
VERSION = '20180901' # Foursquare API version

The function *getNearbyVenues* will be used to explore venues for each neighborhood

In [9]:
def getNearbyVenues(names, latitudes, longitudes, radius, LIMIT):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Let's get the top 100 venues for each neighborhood within a radius of 500 meters.

In [10]:
radius = 500
LIMIT = 100 

ct_venues = getNearbyVenues(names=df_boroughs['Neighborhood'],
                                   latitudes=df_boroughs['Latitude'],
                                   longitudes=df_boroughs['Longitude'],
                                   radius = radius,
                                   LIMIT = LIMIT
                                  )

The Beaches
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West, Lawrence Park
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West, Forest Hill Road Park
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High 

##### Let's check the size of the resulting dataframe

In [11]:
print(ct_venues.shape)
ct_venues.head()

(1624, 7)


Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                            Pantheon       43.677621   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351434   Greek Restaurant

There are total 1624 of venues in the neighborhoods of our dataset. 

The dataframe generated gives interesting informations about venues like their names and their categories (Health Food Store, Pub etc...). This informations are useful to categorize these neigborhoods and get insights about business or real estate investments.

Let's check how many venues were returned for each neighborhood

In [12]:
ct_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            55   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     68   
Christie                                                               16   
Church and Wellesley                                                   75   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                        9   
Dufferin, Dovercourt Village                                           13   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
India Bazaar, The Beaches West                                         19   
Kensington Market, Chinatown, Grange Park                              74   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               45   
Moore Park, Summerhill East                                             2   
North Toronto West, Lawrence Park                                      18   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            33   
Regent Park, Harbourfront                                              44   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     33   
St. James Town                                                         85   
St. James Town, Cabbagetown                                            48   
Stn A PO Boxes                                                         96   
Studio District                                                        37   
Summerhill West, Rathnelly, South Hill, Forest ...                     14   
The Annex, North Midtown, Yorkville                                    19   
The Beaches                                                             4   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             55   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      68   
Christie                                                                16   
Church and Wellesley                                                    75   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              33   
Davisville North                    

Let's find out how many unique categories can be curated from all the returned venues

In [13]:
print('There are {} uniques categories.'.format(len(ct_venues['Venue Category'].unique())))

There are 237 uniques categories.


#### Analyze each neighborhood

We first need to convert the venue categories in one-hot vector form such that if a venue category occurs in a particular neighborhood, we assign 1 to it's corresponding category and 0 where it doesn't occur.

In [14]:
# one hot encoding
ct_onehot = pd.get_dummies(ct_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
ct_onehot['Neighborhood'] = ct_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [ct_onehot.columns[-1]] + list(ct_onehot.columns[:-1])
mct_onehot = ct_onehot[fixed_columns]

ct_onehot.head()

Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       0     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         0   
4        0                0         0             0              0         0   

   Butcher  Café  Cajun / Creole Restaurant  Camera Store  Candy Store  \
0        0     0                          0             0            0   
1        0     0                          0             0            0   
2        0     0                          0             0            0   
3        0     0           

Let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [15]:
ct_grouped = ct_onehot.groupby('Neighborhood').mean().reset_index()
ct_grouped.head()

Neighborhood  Afghan Restaurant  \
0                                        Berczy Park                0.0   
1       Brockton, Parkdale Village, Exhibition Place                0.0   
2  Business reply mail Processing Centre, South C...                0.0   
3  CN Tower, King and Spadina, Railway Lands, Har...                0.0   
4                                 Central Bay Street                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0   0.0000              0.0000        0.0000           0.000            0.000   
1   0.0000              0.0000        0.0000           0.000            0.000   
2   0.0000              0.0000        0.0000           0.000            0.000   
3   0.0625              0.0625        0.0625           0.125            0.125   
4   0.0000              0.0000        0.0000           0.000            0.000   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0            0.0000                  0.0           0.0       0.0     0.018182   
1            0.0000                  0.0           0.0       0.0     0.000000   
2            0.0000                  0.0           0.0       0.0     0.000000   
3            0.0625                  0.0           0.0       0.0     0.000000   
4            0.0000                  0.0           0.0       0.0     0.000000   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0    0.000000                  0.0               0.0                 0.0   
1    0.000000                  0.0               0.0                 0.0   
2    0.000000                  0.0               0.0                 0.0   
3    0.000000                  0.0               0.0                 0.0   
4    0.014706                  0.0               0.0                 0.0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop    Bakery  Bank       Bar  \
0         0.0000        0.0         0.0    0.018182  0.036364   0.0  0.000000   
1         0.0000        0.0         0.0    0.000000  0.043478   0.0  0.043478   
2         0.0625        0.0         0.0    0.000000  0.000000   0.0  0.000000   
3         0.0000        0.0         0.0    0.000000  0.000000   0.0  0.062500   
4         0.0000        0.0         0.0    0.000000  0.000000   0.0  0.000000   

   Baseball Stadium  Basketball Stadium     Beach  Bed & Breakfast  Beer Bar  \
0               0.0            0.018182  0.018182              0.0  0.036364   
1               0.0            0.000000  0.000000              0.0  0.000000   
2               0.0            0.000000  0.000000              0.0  0.000000   
3               0.0            0.000000  0.000000              0.0  0.000000   
4               0.0            0.000000  0.000000              0.0  0.000000   

   Beer Store  Belgian Restaurant    Bistro  Boat or Ferry  Bookstore  \
0         0.0                 0.0  0.018182         0.0000   0.000000   
1         0.0                 0.0  0.000000         0.0000   0.000000   
2         0.0                 0.0  0.000000         0.0000   0.000000   
3         0.0                 0.0  0.000000         0.0625   0.000000   
4         0.0                 0.0  0.000000         0.0000   0.014706   

   Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  \
0    0.0000                   0.0        0.018182   0.0000         0.000000   
1    0.0000                   0.0        0.086957   0.0000         0.000000   
2    0.0000                   0.0        0.000000   0.0625         0.000000   
3    0.0625                   0.0        0.000000   0.0000         0.000000   
4    0.0000                   0.0        0.000000   0.0000         0.029412   

   Building  Burger Joint  Burrito Place  Bus Line  Butcher      Café  \
0       0.0      0.000000       0.000000       0.0   0.0000  0.018182   
1       0.0      0.000000       0.043478       0.0   0.0000  0.130435   
2       0.0      0.000000       0.062500       0.0   0.06

#### Top 10 most common venues in each neighborhood

In [16]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = ct_grouped['Neighborhood']

for ind in np.arange(ct_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(ct_grouped.iloc[ind, :], num_top_venues)


neighborhoods_venues_sorted

Neighborhood 1st Most Common Venue  \
0                                         Berczy Park           Coffee Shop   
1        Brockton, Parkdale Village, Exhibition Place                  Café   
2   Business reply mail Processing Centre, South C...           Pizza Place   
3   CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                  Central Bay Street           Coffee Shop   
5                                            Christie         Grocery Store   
6                                Church and Wellesley           Coffee Shop   
7                      Commerce Court, Victoria Hotel           Coffee Shop   
8                                          Davisville          Dessert Shop   
9                                    Davisville North      Department Store   
10                       Dufferin, Dovercourt Village                Bakery   
11             First Canadian Place, Underground city           Coffee Shop   
12    Forest Hill North & West, Forest Hill Road Park                  Park   
13                           Garden District, Ryerson        Clothing Store   
14  Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
15                      High Park, The Junction South                  Café   
16                     India Bazaar, The Beaches West                  Park   
17          Kensington Market, Chinatown, Grange Park    Mexican Restaurant   
18                                      Lawrence Park                  Park   
19                           Little Portugal, Trinity                   Bar   
20                        Moore Park, Summerhill East            Playground   
21                  North Toronto West, Lawrence Park           Coffee Shop   
22                             Parkdale, Roncesvalles        Breakfast Spot   
23        Queen's Park, Ontario Provincial Government           Coffee Shop   
24                          Regent Park, Harbourfront           Coffee Shop   
25                           Richmond, Adelaide, King           Coffee Shop   
26                                           Rosedale                  Park   
27                                           Roselawn           Music Venue   
28                                 Runnymede, Swansea                  Café   
29                                     St. James Town           Coffee Shop   
30                        St. James Town, Cabbagetown           Coffee Shop   
31                                     Stn A PO Boxes           Coffee Shop   
32                                    Studio District           Coffee Shop   
33  Summerhill West, Rathnelly, South Hill, Forest...           Coffee Shop   
34                The Annex, North Midtown, Yorkville        Sandwich Place   
35                                        The Beaches     Health Food Store   
36                       The Danforth West, Riverdale      Greek Restaurant   
37           Toronto Dominion Centre, Design Exchange           Coffee Shop   
38                     University of Toronto, Harbord                  Café   

            2nd Most Common Venue  3rd Most Common Venue  \
0                    Cocktail Bar     Seafood Restaurant   
1                       Nightclub            Coffee Shop   
2                   Garden Center         Farmers Market   
3                 Airport Service               Boutique   
4                            Café     Italian Restaurant   
5                            Café                   Park   
6             Japanese Restaurant                Gay Bar   
7                      Restaurant                   Café   
8                  Sandwich Place            Pizza Place   
9                         Dog Run                   Park   
10                       Pharmacy            Supermarket   
11                           Café                  Hotel   
12                  Jewelry Store                  Trail   
13                    Coffee Shop       

## Clustering neighborhoods

The k-means model will be used to cluster neighborhoods. The number of clusters is fixed to 4

In [17]:
# set number of clusters
kclusters = 5

ct_grouped_clustering = ct_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(ct_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:35] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 4, 0, 1, 0,
       0, 0, 0, 0, 2, 3, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

Let's create a new dataframe <em>ct_merged</em> that includes the neighborhoods, their clusters and top 10 venues.

In [18]:
# Copy df_boroughs in ct_merged
ct_merged = df_boroughs

# Add clustering labels
ct_merged['Cluster Labels'] = kmeans.labels_

# Merge ct_merged with neighborhoods_venues_sorted
ct_merged = ct_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

ct_merged.head(10)

PostalCode          Borough  \
0        M4E     East Toronto   
1        M4K     East Toronto   
2        M4L     East Toronto   
3        M4M     East Toronto   
4        M4N  Central Toronto   
5        M4P  Central Toronto   
6        M4R  Central Toronto   
7        M4S  Central Toronto   
8        M4T  Central Toronto   
9        M4V  Central Toronto   

                                        Neighborhood   Latitude  Longitude  \
0                                        The Beaches  43.676357 -79.293031   
1                       The Danforth West, Riverdale  43.679557 -79.352188   
2                     India Bazaar, The Beaches West  43.668999 -79.315572   
3                                    Studio District  43.659526 -79.340923   
4                                      Lawrence Park  43.728020 -79.388790   
5                                   Davisville North  43.712751 -79.390197   
6                  North Toronto West, Lawrence Park  43.715383 -79.405678   
7                                         Davisville  43.704324 -79.388790   
8                        Moore Park, Summerhill East  43.689574 -79.383160   
9  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               0     Health Food Store                   Pub   
1               0      Greek Restaurant           Coffee Shop   
2               0                  Park                   Gym   
3               0           Coffee Shop                Bakery   
4               0                  Park              Bus Line   
5               0      Department Store               Dog Run   
6               0           Coffee Shop        Clothing Store   
7               0          Dessert Shop        Sandwich Place   
8               0            Playground                 Trail   
9               0           Coffee Shop           Pizza Place   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                 Trail           Yoga Studio               Dog Run   
1    Italian Restaurant             Bookstore            Restaurant   
2                   Pub        Sandwich Place          Liquor Store   
3             Gastropub               Brewery                  Café   
4           Swim School    Dim Sum Restaurant  Ethiopian Restaurant   
5                  Park  Gym / Fitness Center        Breakfast Spot   
6           Yoga Studio    Chinese Restaurant    Salon / Barbershop   
7           Pizza Place                   Gym    Italian Restaurant   
8           Yoga Studio      Department Store           Escape Room   
9    Light Rail Station          Liquor Store            Restaurant   

  6th Most Common Venue        7th Most Common Venue  \
0    Dim Sum Restaurant                        Diner   
1        Ice Cream Shop       Furniture / Home Store   
2         Burrito Place           Italian Restaurant   
3   American Restaurant            Convenience Store   
4           Escape Room            Electronics Store   
5          Dance Studio                        Hotel   
6  Fast Food Restaurant                          Spa   
7           Coffee Shop                         Café   
8     Electronics Store  Eastern European Restaurant   
9                  Bank                   Bagel Shop   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0               Discount Store   Distribution Center             Donut Shop  
1                          Pub                   Spa      Indian Restaurant  
2                   Restaurant  Fast Food Restaurant      Fish & Chips Shop  
3                    Bookstore    Seafood Restaurant            Cheese Shop  
4  Eastern European Restaurant   Dumpling Restaurant             Donut Shop  
5               Sandwich Place     Food & Drink Shop         Discount Store  
6                        Diner   Sporting Goods Shop             Restaurant  
7             Sushi Restaurant          

#### Now it's time to visualize the clustering with Folium

In [24]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(
        ct_merged['Latitude'], 
        ct_merged['Longitude'], 
        ct_merged['Neighborhood'], 
        ct_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examining clusters

We can examine each cluster and determine the discriminating venue categories that distinguish each cluster.

#### Cluster 1

In [25]:
ct_merged.loc[ct_merged['Cluster Labels'] == 0, ct_merged.columns[[1] + list(range(5, ct_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0       East Toronto               0     Health Food Store   
1       East Toronto               0      Greek Restaurant   
2       East Toronto               0                  Park   
3       East Toronto               0           Coffee Shop   
4    Central Toronto               0                  Park   
5    Central Toronto               0      Department Store   
6    Central Toronto               0           Coffee Shop   
7    Central Toronto               0          Dessert Shop   
8    Central Toronto               0            Playground   
9    Central Toronto               0           Coffee Shop   
10  Downtown Toronto               0                  Park   
11  Downtown Toronto               0           Coffee Shop   
13  Downtown Toronto               0           Coffee Shop   
14  Downtown Toronto               0        Clothing Store   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0           Coffee Shop   
17  Downtown Toronto               0           Coffee Shop   
19  Downtown Toronto               0           Coffee Shop   
21  Downtown Toronto               0           Coffee Shop   
22   Central Toronto               0           Music Venue   
23   Central Toronto               0                  Park   
24   Central Toronto               0        Sandwich Place   
25  Downtown Toronto               0                  Café   
28  Downtown Toronto               0           Coffee Shop   
29  Downtown Toronto               0           Coffee Shop   
30  Downtown Toronto               0         Grocery Store   
31      West Toronto               0                Bakery   
32      West Toronto               0                   Bar   
33      West Toronto               0                  Café   
34      West Toronto               0                  Café   
35      West Toronto               0        Breakfast Spot   
36      West Toronto               0                  Café   
37  Downtown Toronto               0           Coffee Shop   
38      East Toronto               0           Pizza Place   

   2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0                    Pub                  Trail           Yoga Studio   
1            Coffee Shop     Italian Restaurant             Bookstore   
2                    Gym                    Pub        Sandwich Place   
3                 Bakery              Gastropub               Brewery   
4               Bus Line            Swim School    Dim Sum Restaurant   
5                Dog Run                   Park  Gym / Fitness Center   
6         Clothing Store            Yoga Studio    Chinese Restaurant   
7         Sandwich Place            Pizza Place                   Gym   
8                  Trail            Yoga Studio      Department Store   
9            Pizza Place     Light Rail Station          Liquor Store   
10            Playground                  Trail      Department Store   
11                  Café            Pizza Place            Restaurant   
13                  Park                 Bakery                   Pub   
14           Coffee Shop                   Café       Bubble Tea Shop   
15                  Café           Cocktail Bar            Restaurant   
16          Cocktail Bar     Seafood Restaurant                Bakery   
17                  Café     Italian Restaurant        Sandwich Place   
19              Aquarium                  Hotel                  Café   
21            Restaurant                   Café                 Hotel   
22                Garden            Yoga Studio      Department Store   
23         Jewelry Store                  Trail      Sushi Restaurant   
24                  Café            Coffee Shop                  Park   
25             Bookstore                    Bar   Japanese Restaurant   
28    Italian Restaurant                    Pub            Restaurant   
29               

#### Cluster 2

In [21]:
ct_merged.loc[ct_merged['Cluster Labels'] == 1, ct_merged.columns[[1] + list(range(5, ct_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
20  Downtown Toronto               1           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
20                 Hotel                  Café            Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
20           Salad Place   American Restaurant    Seafood Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
20   Japanese Restaurant      Asian Restaurant             Steakhouse

#### Cluster 3

In [22]:
ct_merged.loc[ct_merged['Cluster Labels'] == 2, ct_merged.columns[[1] + list(range(5, ct_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
12  Downtown Toronto               2           Coffee Shop   
26  Downtown Toronto               2    Mexican Restaurant   

            2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
12            Japanese Restaurant               Gay Bar      Sushi Restaurant   
26  Vegetarian / Vegan Restaurant           Coffee Shop                  Café   

   5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
12            Restaurant                    Pub           Men's Store   
26                   Bar  Vietnamese Restaurant           Gaming Cafe   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
12  Mediterranean Restaurant                 Hotel            Yoga Studio  
26                      Park          Dessert Shop                 Bakery

#### Cluster 4

In [23]:
ct_merged.loc[ct_merged['Cluster Labels'] == 3, ct_merged.columns[[1] + list(range(5, ct_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
27  Downtown Toronto               3        Airport Lounge   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
27       Airport Service              Boutique                 Plane   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
27               Airport    Airport Food Court          Airport Gate   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
27      Airport Terminal                   Bar        Harbor / Marina

#### CLuster 5

In [27]:
ct_merged.loc[ct_merged['Cluster Labels'] == 4, ct_merged.columns[[1] + list(range(5, ct_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18  Downtown Toronto               4           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18                  Café                   Gym                 Hotel   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
18                   Bar            Restaurant        Clothing Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
18       Thai Restaurant         Deli / Bodega             Steakhouse